### PR TITLE
Implement with instance chains

### PR DESCRIPTION
**Pull request by garyb**
I think as per #9 we probably want to deprecate this library and put these classes with `Either`/`Coproduct`, but just thought I'd try this out.

@LiamGoodacre I was a little surprised to find that it didn't raise an error even when these were defined without instance chains until used. I thought purescript/purescript#3129 would have changed that? Maybe I misunderstood.

### CODE REVIEW

**Comment by LiamGoodacre** [comment]
Ooooh, I will have a play tomorrow and see if I messed something up.  Thanks!

File: src/Data/Inject/Coproduct.purs [hunk 1]
```diff
@@ -0,0 +1,23 @@
+module Data.Inject.Either where
```
Comment (LiamGoodacre) [inline src/Data/Inject/Coproduct.purs]: Should be `Data.Inject.Coproduct`?

**Review by LiamGoodacre — CHANGES_REQUESTED** [review]

File: src/Data/Inject/Coproduct.purs (hunk 1 above)
Comment (garyb) [reply src/Data/Inject/Coproduct.purs]: Uh yes :wink:

**Comment by garyb** [comment]
Ack, maybe it was user error on my part then... I did pull master and `stack install` for sure, but maybe I still did something wrong in there / it was using the wrong PATH or something.

The other thing I thought of, is this version using instance chains can inject into the left or the right, the original definition is biased to one side, so perhaps I tested with that and it failed. I'll try and reproduce again later today, sorry to waste your time!

**Comment by garyb** [comment]
@LiamGoodacre I figured it out. I did `git pull master` but not `git pull purescript master` 🤦‍♂️

File: src/Data/Inject/Coproduct.purs [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+instance injectLeft :: Inject f g => Inject f (Coproduct g h) where
+  inj = Coproduct <<< Left <<< inj
+  prj = coproduct prj (const Nothing)
+
+else instance injectRight :: Inject f g => Inject f (Coproduct h g) where
+  inj = Coproduct <<< Right <<< inj
+  prj = coproduct (const Nothing) prj
+
+else instance injectReflexive :: Inject f f where
+  inj = id
```
Comment (chexxor) [inline src/Data/Inject/Coproduct.purs]: id => identity

File: src/Data/Inject/Coproduct.purs (hunk 1 above)
Comment (chexxor) [reply src/Data/Inject/Coproduct.purs]: Yeah, this change is required to compile - it collides with the other Data.Inject.Either module in this project.

File: src/Data/Inject/Either.purs [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+instance injectLeft :: Inject a b => Inject a (Either b c) where
+  inj = Left <<< inj
+  prj = either prj (const Nothing)
+
+else instance injectRight :: Inject a b => Inject a (Either c b) where
+  inj = Right <<< inj
+  prj = either (const Nothing) prj
+
+else instance injectReflexive :: Inject a a where
+  inj = id
```
Comment (chexxor) [inline src/Data/Inject/Either.purs]: id => identity

**Review by chexxor — COMMENTED** [review]
Requires a few changes to compile with the 0.12 package-set.

**Comment by chexxor** [comment]
Requested changes are addressed in https://github.com/purescript/purescript-inject/pull/11, so you can merge that, then merge this PR later.

**Comment by LiamGoodacre** [comment]
Closing as we will move these type classes and instances to functors/either.

**Pull request closed without merging**